rllib/algorithms/dqn/tf/dqn_tf_learner.py

```diff
@@ -0,0 +1,100 @@
+from typing import Any, Dict, Mapping
+from ray.rllib.algorithms.dqn.dqn_learner import QF_PREDS, QF_TARGET_PREDS
+from ray.rllib.core.learner.learner import LearnerHyperparameters
+from ray.rllib.core.learner.tf.tf_learner import TfLearner
+from ray.rllib.core.rl_module.rl_module import ModuleID
+from ray.rllib.execution.common import (
+    LAST_TARGET_UPDATE_TS,
+    NUM_TARGET_UPDATES,
+)
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.framework import try_import_tf
+from ray.rllib.utils.nested_dict import NestedDict
+from ray.rllib.utils.typing import TensorType
+
+_, tf, _ = try_import_tf()
+
+
+class DQNTfLearner(TfLearner):
+    """Implements loss and update logic for DQN."""
+
+    @override(TfLearner)
+    def compute_loss_for_module(
+        self,
+        *,
+        module_id: ModuleID,
+        hps: LearnerHyperparameters,
+        batch: NestedDict,
+        fwd_out: Mapping[str, TensorType]
+    ) -> TensorType:
+        # The DQN Module returns ACTIONS for the maximum
+        # Q values.
+        q_values_t = fwd_out[QF_PREDS]
+        q_target_values_tp1 = fwd_out[QF_TARGET_PREDS]
+
+        terminated_mask = tf.cast(batch[SampleBatch.TERMINATEDS], tf.float32)
+
+        # Create a mask for non-terminal experiences.
+        q_target_values_tp1 = (1.0 - terminated_mask) * q_target_values_tp1
+        # Compute right-hand side of Bellman equation to compute training targets.
+        # TODO (simon): Make it n-step. This means also that the `TargetEncoder`
+        # has to account for this. Maybe another solution is better here.
+        q_t_targets = batch[SampleBatch.REWARDS] + hps.gamma * q_target_values_tp1
+        # Compute TD error.
+        # TODO (simon): Check, if `stop_gradient()` is even needed when
+        # `trainable` is set to `False` in the `DQNRLModule`.
+        td_error = q_values_t - tf.stop_gradient(q_t_targets)
+        # TODO (simon): If recurrent network is used we must use a mask.
+        total_loss = tf.reduce_mean(
+            tf.cast(batch[SampleBatch.PRIO_WEIGHTS], tf.float32)
+            * hps.td_error_loss_fn(td_error)
+        )
+
+        self.register_metrics(
+            {
+                "total_loss": total_loss,
+                "mean_q": tf.reduce_mean(q_values_t),
+                "min_q": tf.reduce_min(q_values_t),
+                "max_q": tf.reduce_max(q_values_t),
+                "mean_td_error": tf.reduce_mean(td_error),
+            }
+        )
+
+        return total_loss
+
+    @override(TfLearner)
+    def additional_update_for_module(
+        self,
+        *,
+        module_id: ModuleID,
+        hps: LearnerHyperparameters,
+        timestep: int,
+        last_update: int,
+        **kwargs
+    ) -> Dict[str, Any]:
+        results = super().additional_update_for_module(
```

Comment:
in appo learner we take this whole additional update and make it apart of the base class, in this case that would be the DQNLearner (which doesnt exist yet, but should since itll be useful when making the torch learner)

then you can have a function called in the base function where you call a target network update function. The target network update is done with tau averaging (like polyak averaging except polyak = 1 - tau)


see this line of code over here, which does what im describing. the function is implemented in the appo torch and tf learners. 

https://github.com/avnishn/ray/blob/f197fbca1c59db69890d7b66eeb5ca14862fe76b/rllib/algorithms/appo/appo_learner.py#L98